include/DDpackage.hpp

```
@@ -0,0 +1,2047 @@
+                    }
+                    rootweight.from_string(m.str(1), m.str(2));
+                }
+
+                while (std::getline(is, line)) {
+                    if (line.empty() || line.size() == 1) continue;
+
+                    if (!std::regex_match(line, m, line_regex)) {
+                        throw std::runtime_error("Regex did not match line: " + line);
+                    }
+
+                    // match 1: node_idx
+                    // match 2: qubit_idx
+
+                    // repeats for every edge
+                    // match 3: edge content
+                    // match 4: edge_target_idx
+                    // match 5: real + imag (without i)
+                    // match 6: real
+                    // match 7: imag (without i)
+                    node_index = std::stoi(m.str(1));
+                    v          = static_cast<Qubit>(std::stoi(m.str(2)));
+
+                    for (auto edge_idx = 3U, i = 0U; i < N; i++, edge_idx += 5) {
+                        if (m.str(edge_idx).empty()) continue;
+
+                        edge_indices[i] = std::stoi(m.str(edge_idx + 1));
+                        edge_weights[i].from_string(m.str(edge_idx + 3), m.str(edge_idx + 4));
+                    }
+
+                    result = deserializeNode(node_index, v, edge_indices, edge_weights, nodes);
+                }
+            }
+
+            auto w = cn.getCached(rootweight.r, rootweight.i);
+            ComplexNumbers::mul(w, result.w, w);
+            result.w = cn.lookup(w);
+            cn.returnToCache(w);
+
+            return result;
+        }
+
+        template<class Node, class Edge = Edge<Node>>
+        Edge deserialize(const std::string& inputFilename, bool readBinary) {
+            auto ifs = std::ifstream(inputFilename);
+
+            if (!ifs.good()) {
+                std::cerr << "Cannot open serialized file: " << inputFilename << std::endl;
```

I'd rather the app throws/crashes early instead of returning a default.

---

Agreed. Now throws an `std::invalid_argument`.